### ISSUE TITLE
[critical] POST requests with JSON data using a {{variable}} set to undefined or null fails with error: Cannot read property 'toString' of undefined

### ISSUE THREAD

**Issue by aksbenz** [postmanlabs/newman#1381]
Newman exists when a variable is being used in the request body, but the variable was set to undefined or null, as it was not available in the previous request where it was being set.

Rather than passing the variable as it is and the tests failing, newman exits with error and no reports are created.

Error shown by newman:
**Cannot read property 'toString' of undefined**

 1. Newman Version (can be found via `newman -v`): 3.9.2
 2. OS details (type, version, and architecture): Windows 7 Pro, x64
 3. Are you using Newman as a library, or via the CLI? CLI
 3. Did you encounter this recently, or has this bug always been there: Probably after 3.8.2
 4. Expected behaviour: Any undefined or null variables should be resolved to undefined/null and sent accross
 5. Command / script used to run Newman: 
 6. Sample collection, and auxiliary files (minus the sensitive details): Pasted at the end
 7. Screenshots (if applicable):

Steps to reproduce the problem:

1. Create a POST Request with JSON body
2. In Pre-requisite or a request before, Set the variable to undefined. For example: 
```
pm.environment.set("version", undefined);
```
2. In JSON body use a variable, for example : 
```
{"version":"{{version}}"}
```
3. Run the collection using newman

Newman CLI shows following output:
Temp

→ After comparisons
  GET https://postman-echo.com/time/after?timestamp={{start}}&target=2017-10-10 [400 Bad Request, 408B, 1622ms]

→ POST Raw Text

┌─────────────────────────┬──────────┬──────────┐
│                         │ executed │   failed │
├─────────────────────────┼──────────┼──────────┤
│              iterations │        0 │        0 │
├─────────────────────────┼──────────┼──────────┤
│                requests │        1 │        0 │
├─────────────────────────┼──────────┼──────────┤
│            test-scripts │        1 │        0 │
├─────────────────────────┼──────────┼──────────┤
│      prerequest-scripts │        1 │        0 │
├─────────────────────────┼──────────┼──────────┤
│              assertions │        0 │        0 │
├─────────────────────────┴──────────┴──────────┤
│ total run duration: 0ms                       │
├───────────────────────────────────────────────┤
│ total data received: 0B (approx)              │
├───────────────────────────────────────────────┤
│ average response time: 1622ms                 │
└───────────────────────────────────────────────┘
**Cannot read property 'toString' of undefined**

Postman Collection Runner gives following error stack:
message
:
"Cannot read property 'toString' of undefined"
name
:
"TypeError"
stack
:
"TypeError: Cannot read property 'toString' of undefined↵    at PostmanVariable.toString (C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\lib\collection\variable.js:108:34)↵    at C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\lib\superstring\index.js:221:53↵    at C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\lib\superstring\index.js:47:23↵    at RegExp.[Symbol.replace] (<anonymous>)↵    at String.replace (native)↵    at SuperString.replace (C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\lib\superstring\index.js:45:34)↵    at Substitutor.parse (C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\lib\superstring\index.js:132:27)↵    at Function.replaceSubstitutions (C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\lib\collection\property.js:222:69)↵    at Function.<anonymous> (C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\lib\collection\property.js:253:25)↵    at C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\lib\collection\property.js:247:40↵    at C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\node_modules\lodash\lodash.js:4944:15↵    at baseForOwn (C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\node_modules\lodash\lodash.js:3001:24)↵    at Function.forOwn (C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\node_modules\lodash\lodash.js:13002:24)↵    at Function.<anonymous> (C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\lib\collection\property.js:246:19)↵    at C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\lib\collection\property.js:247:40↵    at C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\node_modules\lodash\lodash.js:4944:15↵    at baseForOwn (C:\Users\aadm221\AppData\Local\Postman\app-5.5.0\resources\app.asar\node_modules\postman-collection\node_modules\lodash\lodash.js:3001:24)"

Sample Collection:
```
{
	"info": {
		"name": "Temp",
		"_postman_id": "734b99af-72b1-a510-dc4c-10214c91aeee",
		"description": "",
		"schema": "https://schema.getpostman.com/json/collection/v2.0.0/collection.json"
	},
	"item": [
		{
			"name": "After comparisons",
			"event": [
				{
					"listen": "test",
					"script": {
						"id": "b8d9c216-89af-4c77-8ff6-7066d5e8fd77",
						"type": "text/javascript",
						"exec": [
							"pm.environment.set(\"version\", undefined);"
						]
					}
				}
			],
			"request": {
				"method": "GET",
				"header": [],
				"body": {},
				"url": {
					"raw": "https://postman-echo.com/time/after?timestamp={{start}}&target=2017-10-10",
					"protocol": "https",
					"host": [
						"postman-echo",
						"com"
					],
					"path": [
						"time",
						"after"
					],
					"query": [
						{
							"key": "timestamp",
							"value": "{{start}}",
							"equals": true
						},
						{
							"key": "target",
							"value": "2017-10-10",
							"equals": true
						}
					]
				},
				"description": "A simple `GET` request to `/time/after` to check if the provided timestamps is after a comparison `target` (query parameter).\n\nThis endpoint accepts `timestamp`, `locale`, `format`, `strict`, and `target` query parameters to construct the date time instance to check against.\n\nResponses are provided in JSON format, with a `after` key to indicate the result. The response code is `200` for valid query parameters, and `400` otherwise.\n\n```\n{\n  after: true/false\n}\n```"
			},
			"response": []
		},
		{
			"name": "POST Raw Text",
			"event": [
				{
					"listen": "test",
					"script": {
						"id": "eca79d5c-0416-4fc6-b4b6-eafe6f371f05",
						"type": "text/javascript",
						"exec": [
							"var responseJSON;",
							"",
							"try { ",
							"    responseJSON = JSON.parse(responseBody); ",
							"    tests['response is valid JSON'] = true;",
							"}",
							"catch (e) { ",
							"    responseJSON = {}; ",
							"    tests['response is valid JSON'] = false;",
							"}",
							"",
							"",
							"tests['response has post data'] = _.has(responseJSON, 'data');",
							"tests['response matches the data posted'] = (responseJSON.data && responseJSON.data.length === 256);",
							"",
							"tests[\"content-type equals text/plain\"] = responseJSON && responseJSON.headers && (responseJSON.headers[\"content-type\"] === 'text/plain');"
						]
					}
				},
				{
					"listen": "prerequest",
					"script": {
						"id": "c5d968db-e436-46e3-b4fe-cda9d6ebab89",
						"type": "text/javascript",
						"exec": [
							""
						]
					}
				}
			],
			"request": {
				"method": "POST",
				"header": [
					{
						"key": "Content-Type",
						"value": "application/json"
					}
				],
				"body": {
					"mode": "raw",
					"raw": "{\n\t\"name\":\"postman\",\n\t\"value\":\"{{version}}\"\n}"
				},
				"url": "https://postman-echo.com/post",
				"description": "The HTTP `POST` request method is meant to transfer data to a server \n(and elicit a response). What data is returned depends on the implementation\nof the server.\n\nA `POST` request can pass parameters to the server using \"Query String \nParameters\", as well as the Request Body. For example, in the following request,\n\n> POST /hi/there?hand=wave\n>\n> <request-body>\n\nThe parameter \"hand\" has the value \"wave\". The request body can be in multiple\nformats. These formats are defined by the MIME type of the request. The MIME \nType can be set using the ``Content-Type`` HTTP header. The most commonly used \nMIME types are:\n\n* `multipart/form-data`\n* `application/x-www-form-urlencoded`\n* `application/json`\n\nThis endpoint echoes the HTTP headers, request parameters, the contents of\nthe request body and the complete URI requested."
			},
			"response": []
		}
	]
}
```

**Comment by kunagpal** [comment]
@aksbenz While an error being thrown is indeed unacceptable here, `undefined` is not a part of the JSON standard, which is why calling `JSON.stringify` with any `undefined` values will result in them being skipped from the result altogether. `null`, on the other hand, is completely valid, and thus could be considered here as an exception. Out of curiosity, is there a specific reason for sending an `undefined` value as a part of the JSON body?

**Comment by aksbenz** [comment]
Not intentionally sending undefined.

There are a lot of cases, where some data is set from the response of a previous request , but in case no data is received for that field, it gets set to undefined/null. Don't want to put all kinds of null checks in these requests as they are used for fetching data.

So the later request which is using this variable would eventually fail, since the response would give error.

But I do not expect newman to completely exit with no reports and without finishing with all the tests.

**Comment by kunagpal** [comment]
@aksbenz This has been fixed in https://github.com/postmanlabs/postman-collection/pull/539, will go out with the next Newman release. 😄

**Comment by aksbenz** [comment]
Thanks @kunagpal

**Comment by kunagpal** [comment]
@aksbenz This has been fixed with Newman v3.9.3. 😄
